cohort-engine/src/main/java/com/ibm/cohort/engine/terminology/R4RestFhirTerminologyProvider.java

```diff
@@ -0,0 +1,118 @@
+	 * in the original OSS implementation and just silently ignored. I added logic
+	 * that will throw UnsupportedOperationException if either is encountered during
+	 * execution.</li>
+	 * <li>If an incorrect number of ValueSet resources are returned in the search
+	 * by URL operation, the OSS would either pass it through silently as 0 codes
+	 * when no records were found or arbitrarily pick the first result row in the
+	 * case of multiple results. Those conditions will now yield an
+	 * IllegalArgumentException.</li>
+	 * </ol>
+	 */
+	@Override
+    public Boolean resolveByUrl(ValueSetInfo valueSet) {
+        if (valueSet.getVersion() != null
+                || (valueSet.getCodeSystems() != null && valueSet.getCodeSystems().size() > 0)) {
+            if (!(valueSet.getCodeSystems().size() == 1 && valueSet.getCodeSystems().get(0).getVersion() == null)) {
+                throw new UnsupportedOperationException(String.format(
+                        "Could not expand value set %s; version and code system bindings are not supported at this time.",
+                        valueSet.getId()));
+            }
+        }
+        
+        if (valueSet.getId().startsWith("urn:oid:")) {
+            valueSet.setId(valueSet.getId().replace("urn:oid:", ""));
+        } else if (valueSet.getId().startsWith("http:") || valueSet.getId().startsWith("https:")) {
+            Bundle searchResults = fhirClient.search().forResource(ValueSet.class)
+                    .where(ValueSet.URL.matches().value(valueSet.getId())).returnBundle(Bundle.class).execute();
+            if (searchResults.isEmpty()) {
+                throw new IllegalArgumentException(String.format("Could not resolve value set %s.", valueSet.getId()));
+            } else if (searchResults.getEntry().size() == 1) {
+                valueSet.setId(searchResults.getEntryFirstRep().getResource().getIdElement().getIdPart());
+            } else {
+                throw new IllegalArgumentException("Found more than 1 ValueSet with url: " + valueSet.getId());
+            }
+        }
+
+        return true;
+    }
+	
+	/** 
+	 * This is a small patch to the OSS implementation to use named-parameter lookup on 
+	 * the operation response instead of just assuming a positional location.
+	 */
+    @Override
+    public Code lookup(Code code, CodeSystemInfo codeSystem) throws ResourceNotFoundException {
+        Parameters respParam = fhirClient
+                .operation()
+                .onType(CodeSystem.class)
+                .named("lookup")
+                .withParameter(Parameters.class, "code", new CodeType(code.getCode()))
+                .andParameter("system", new UriType(codeSystem.getId()))
+                .execute();
+        
+        StringType display = (StringType) respParam.getParameter("display");
+
+        return code.withSystem(codeSystem.getId())
+                .withDisplay(display != null ? display.getValue() : null);
```

Comment:
In the original code, this would do a lookup by positional index and then either 1) properly or improperly use whatever happened to be at that index or 2) throw an NPE if their wasn't a parameter provided. My intention here was to adjust so that point 1 became more deterministic, avoid point2, and alter as few lines as possible from the original implementation.  I agree, however, that it is possible that this new implementation will overwrite the display from the input if it is there and we probably shouldn't do that. Will change it.